clearwater-queue-manager.root/usr/share/clearwater/clearwater-queue-manager/scripts/upgrade_cassandra.py

```diff
@@ -0,0 +1,204 @@
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or (at your
+# option) any later version, along with the "Special Exception" for use of
+# the program along with SSL, set forth below. This program is distributed
+# in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+from metaswitch.clearwater.etcd_shared.plugin_utils import run_command, safely_write
+import logging
+import os
+import re
+import shutil
+import subprocess
+
+# Not sure if we need the non-CC packages here
+PACKAGES = "cassandra clearwater-cassandra homestead-cassandra homestead-prov-cassandra homer-cassandra memento-cassandra"
+SOURCES_FILE = "/etc/apt/sources.list.d/clearwater.list"
+TEMP_SOURCES_FILE = "/tmp/clearwater.list"
+VERSION_FILE = "/etc/clearwater/cassandra_version"
+
+_log = logging.getLogger("cassandra_upgrade_plugin")
+
+def get_cassandra_version():
+    version = ""
+    nodetool_process = subprocess.Popen("/usr/share/clearwater/bin/run-in-signaling-namespace nodetool version",
+                                        stdout=subprocess.PIPE,
+                                        shell=True)
+
+    # The output of nodetool version is like "ReleaseVersion: 2.0.14"
+    version_regex = "ReleaseVersion: (.*)$"
+    for line in nodetool_process.stdout:
+        line = line.strip()
+        m = re.match(version_regex, line)
+        if m:
+            version = m.group(1)
+            break
+
+    return version
+
+# Attempts the Cassandra upgrade.
+# Returns True if the upgrade succeeded or no upgrade was necessary.
```

Comment:
There's a lot of different response situations to cover with just True/False - I recommend a return code instead